### PR TITLE
Shell: Various fixes for POSIX boogs

### PR DESCRIPTION
**Pull request by timschumi**
This gets us a bit further in LLVMs `config.guess`.

### CODE REVIEW

**Comment by timschumi** [comment]
Something is still not quite right about the newline handling after heredocs, making this a draft for now.

**Review by konradekk — COMMENTED** [review]
(looks like `initialize` lambda has grown – thought itʼd be nice to extract it to a separate method if possible; still, thatʼs rather out of scope, so ignore this suggestion if not interested!)

File: Userland/Shell/main.cpp [hunk 1]
```diff
@@ -171,7 +171,7 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     StringView format;
     bool should_format_live = false;
     bool keep_open = false;
-    bool posix_mode = false;
+    bool posix_mode = (LexicalPath::basename(arguments.strings[0]) == "sh"sv);
```
Comment (alimpfard) [inline Userland/Shell/main.cpp:174]: Extra parens: intentional?

**Review by alimpfard — COMMENTED** [review]
What's wrong with heredocs (after your fix in this PR)?

**Comment by timschumi** [comment]
Without 644acb089cda4f1e45deb00fafcc2675fd86e09d, this fails:

```
#!/bin/sh

if true; then
  cat << EOF
whf
EOF
fi
```

```
Shell Syntax Error: Expected 'else' or 'fi', not Eof
  0| #!/bin/sh
~~~~~
  1| 
```

That seems to happen because ending the heredoc consumes the following newline, so our parser doesn't recognize `fi` as the start of a command, so it never parses it as a keyword.

Without the part that consumes the newline later (in `handle_heredoc_contents`) the following fails with the following message (if I recall correctly I tracked it down to it finding the newline at a wrong point in time and it assuming that the command list is done):

```
#!/bin/sh

cat << EOF
whf
EOF
echo "asdasd"
```

```
Shell Syntax Error: Extra tokens after complete command
  2| cat << EOF
  3| whf
  4| EOF
~~~~~
  5| echo "asdasd"
```

With that patch (as seen in the PR) applied, the it instead fails because it just seems to interpret `echo "asdasd"` as part of the heredoc and tries to open files with those names.

With the "newline consumer" part moved into the while loop above (where it really should be, since we are otherwise consuming newlines every time when we call `handle_heredoc_contents`, even if there is no heredoc content), it still fails with the exact same message (and it also seems to not append the newline _before_ the `EOF` key to the heredoc content).

File: Userland/Shell/main.cpp (hunk 1 above)
Comment (timschumi) [reply Userland/Shell/main.cpp:174]: Kind of, I dislike the less-readability of the non-parens one.

**Comment by timschumi** [comment]
I force-pushed away the patch in question, that requires quite a bit more time from my side and/or another set of eyes that is more familiar with the issue at hand. The other patches should be fine as far as I can tell.

**Pull request merged**